<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=Edge">

<link href="../../cowidget/CoWidget.css" rel="stylesheet" type="text/css" />

<script type="text/javascript">
	'use strict';
	var coWidgetConfig = {
		isDebug : true
	};
</script>
<script type="text/javascript" src="../../cowidget/CoWidget.js"></script>


<title>Example</title>
<body class="api">


	<div>async function</div>

	<br />
	<div class="logger">
		<code id="logger"></code>
	</div>

	<script type="text/javascript">
	'use strict';
	let LOG = cowidget.common.LogFactory.getLog();
	
	class Fly {
		constructor(options) {
			Object.assign(this, {options: options});
		}
	}
	
	class xPromise{
		
		constructor(options) {
			//this._init(options);
		}
		
		async placeAt(options) {
			let promise = new Promise((resolve, reject) => {		
				if(true){
					let timeout = Math.floor(Math.random() * Math.floor(5000));
					LOG.debug('timeout: ', timeout);
					setTimeout(function () {
					    	LOG.debug('resolve: setTimeout options: ', options);
		
							let fly = new Fly(options);					
							resolve(fly);
							//reject('Error');
					    }, timeout);
				}else {
					let xhrProps = {
							url: '../cowidget/CoWidget.js',
							sync: true,
							handleAs: 'text',
							
							load: (data) => {
						    	//LOG.debug('resolve: ', data);
								resolve(data);
							}
					}
					
					cowidget.common.NetXhr.xhr(xhrProps);
				}
			});
			
			promise.then(function (value) {
				  LOG.debug('then: ' + value);
				  return value;
				}).catch(function (error) {
				  LOG.debug(error);
				});
		
			//Object.assign(this, {promise: promise});

			let result = await Promise.resolve(promise);
			LOG.debug('_init: ');
			//LOG.debug('', result);
			Object.assign(this, {fly: result});
			try {
				await Promise.resolve(this.promise);
				LOG.debug('placeAt: ');
				//LOG.debug('', this.fly);
			}catch(exception){
				LOG.error('exception: ', exception);
			}
			
			LOG.debug('placeAt result: ', result);
			return this;
		}
		
		fork() {
			
			try {
				//await Promise.resolve(this.promise);
				LOG.debug('fork: ', this.fly);
			}catch(exception){
				LOG.error('exception: ', exception);
			}
			
			return this;
		}
	}
		
	new xPromise().placeAt({fork: 'A'});
 	let b = new xPromise();
 	let c = new xPromise();
	
	c.placeAt({fork: 'B'});
 	b.placeAt({fork: 'C'});
	
	
// 	try {
// 		// not work
// 		//c.placeAt().fork();
// 		LOG.debug('c.placeAt();', c.placeAt());
// 	}catch(exception){
// 		console.error('exception: ', exception);
// 	}
	
	//LOG.debug('top: ', b);
	//LOG.debug('top: ', c);
	LOG.debug('end')
	</script>
</body>
</html>